cmd/zpool/zpool_main.c

```diff
@@ -2402,226 +2433,260 @@ print_status_config(zpool_handle_t *zhp, status_cbdata_t *cb, const char *name,
 			state = gettext("AVAIL");
 	}
 
-	printf_color(health_str_to_color(state),
-	    "\t%*s%-*s  %-8s", depth, "", cb->cb_namewidth - depth,
-	    name, state);
-
-	if (!isspare) {
-		if (vs->vs_read_errors)
-			rcolor = ANSI_RED;
-
-		if (vs->vs_write_errors)
-			wcolor = ANSI_RED;
-
-		if (vs->vs_checksum_errors)
-			ccolor = ANSI_RED;
-
-		if (cb->cb_literal) {
-			fputc(' ', stdout);
-			printf_color(rcolor, "%5llu",
-			    (u_longlong_t)vs->vs_read_errors);
-			fputc(' ', stdout);
-			printf_color(wcolor, "%5llu",
-			    (u_longlong_t)vs->vs_write_errors);
-			fputc(' ', stdout);
-			printf_color(ccolor, "%5llu",
-			    (u_longlong_t)vs->vs_checksum_errors);
-		} else {
-			zfs_nicenum(vs->vs_read_errors, rbuf, sizeof (rbuf));
-			zfs_nicenum(vs->vs_write_errors, wbuf, sizeof (wbuf));
-			zfs_nicenum(vs->vs_checksum_errors, cbuf,
-			    sizeof (cbuf));
-			fputc(' ', stdout);
-			printf_color(rcolor, "%5s", rbuf);
-			fputc(' ', stdout);
-			printf_color(wcolor, "%5s", wbuf);
-			fputc(' ', stdout);
-			printf_color(ccolor, "%5s", cbuf);
-		}
-		if (cb->cb_print_slow_ios) {
-			if (children == 0)  {
-				/* Only leafs vdevs have slow IOs */
-				zfs_nicenum(vs->vs_slow_ios, rbuf,
-				    sizeof (rbuf));
+	/* Print columns 1-2, vdev name, state */
+	if (cb->cb_print_unhealthy) {
+		if (children > 0) {
+		/* If check_child_health() is true, a child is unhealthy */
+			if (check_child_health(nv))
+				print_parent = B_TRUE;
+		}
+	}
+	if (! (depth > 0 &&
+	    cb->cb_print_unhealthy &&
+	    vs->vs_state == VDEV_STATE_HEALTHY &&
+	    vs->vs_read_errors == 0 &&
+	    vs->vs_write_errors == 0 &&
+	    vs->vs_checksum_errors == 0 &&
+	    print_parent == B_FALSE)) {
```

Comment:
I like the approach of pruning entirely healthy branches, but it looks to me like we could significantly simply this with a few minor modifications.

1) Since `print_status_config()` is called recursively you can optimize this by returning immediately when encountering any branch of the vdev tree with no errors.  This way you also don't need to pull all of the following code under a conditional.  

2) You should also be able to use the `for_each_vdev_in_nvlist()` helper function to iterate over all of the leaf vdevs under a branch.  It'll handle to recursion for you and OR all the return values together.  This is nice because then you just need to return 1 for any leaf vdev matching the unhealthy criteria.

It needs some polish but I'm thinking something like this.  There might be some gotchas with special, log, and l2arc devices so I'd  suggest doing some testing these just to make sure.

```c
static int
is_vdev_healthy(void *hdl_data, nvlist_t *nv, void *data)
{
       status_cbdata_t *cb = data;
       vdev_stat_t *vs;
       uint_t vsc;

       if (nvlist_lookup_uint64_array(nv, ZPOOL_CONFIG_VDEV_STATS,
           (uint64_t **)&vs, &vsc) != 0)
               return (1);

      /* Check state, errors, and return 1 on unhealthy */
      ...
      
      return (0);
}

print_status_config()
{
       ....

       if (cb->cb_print_unhealthy && depth > 0 &&
           for_each_vdev_in_nvlist(nv, is_vdev_healthy, cb) == 0) {
               return;
       }
       
       ...
} 
```